File: leetcode/1032-stream-of-characters/Solution.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayDeque;
import java.util.Deque;

class StreamChecker {
    private static class TrieNode {
        Map<Character, TrieNode> children = new HashMap<>();
        boolean isWord = false;
    }

    private final TrieNode root = new TrieNode();
    private final Deque<Character> stream = new ArrayDeque<>();
    private int maxLen = 0;

    public StreamChecker(String[] words) {
        for (String w : words) {
            maxLen = Math.max(maxLen, w.length());
            TrieNode node = root;
            for (int i = w.length() - 1; i >= 0; i--) {
                char ch = w.charAt(i);
                node = node.children.computeIfAbsent(ch, c -> new TrieNode());
            }
            node.isWord = true;
        }
    }

    public boolean query(char letter) {
        stream.addLast(letter);
        while (stream.size() > maxLen) stream.removeFirst();

        TrieNode node = root;
        for (char ch : reverseView()) {
            TrieNode next = node.children.get(ch);
            if (next == null) return false;
            node = next;
            if (node.isWord) return true;
        }
        return false;
    }

    private Iterable<Character> reverseView() {
        java.util.List<Character> list = new java.util.ArrayList<>(stream);
        java.util.Collections.reverse(list);
        return list;
    }
}

File: leetcode/1032-stream-of-characters/solution.cpp
#include <vector>
#include <string>
#include <deque>
#include <unordered_map>
using namespace std;

class StreamChecker {
    struct TrieNode {
        unordered_map<char, TrieNode*> children;
        bool isWord = false;
    };

    TrieNode* root;
    deque<char> stream;
    int maxLen = 0;

public:
    StreamChecker(vector<string>& words) {
        root = new TrieNode();
        for (const string& w : words) {
            maxLen = max(maxLen, (int)w.size());
            TrieNode* node = root;
            for (int i = (int)w.size() - 1; i >= 0; i--) {
                char ch = w[i];
                if (node->children.find(ch) == node->children.end()) {
                    node->children[ch] = new TrieNode();
                }
                node = node->children[ch];
            }
            node->isWord = true;
        }
    }

    bool query(char letter) {
        stream.push_back(letter);
        while ((int)stream.size() > maxLen) stream.pop_front();

        TrieNode* node = root;
        for (auto it = stream.rbegin(); it != stream.rend(); ++it) {
            auto found = node->children.find(*it);
            if (found == node->children.end()) return false;
            node = found->second;
            if (node->isWord) return true;
        }
        return false;
    }
};
